lock-api/src/main/java/com/palantir/lock/client/MultiClientTransactionStarter.java

```diff
@@ -0,0 +1,323 @@
+        @Value.Parameter
+        Namespace namespace();
+
+        @Value.Parameter
+        RequestParams params();
+
+        static NamespaceAndRequestParams of(Namespace namespace, RequestParams params) {
+            return ImmutableNamespaceAndRequestParams.of(namespace, params);
+        }
+    }
+
+    @Value.Immutable
+    interface RequestParams {
+        @Value.Parameter
+        Integer numTransactions();
+
+        @Value.Parameter
+        StartTransactionsLockWatchEventCache cache();
+
+        @Value.Parameter
+        LockCleanupService lockCleanupService();
+
+        static RequestParams of(
+                int numTransactions,
+                StartTransactionsLockWatchEventCache cache,
+                LockCleanupService lockCleanupService) {
+            return ImmutableRequestParams.of(numTransactions, cache, lockCleanupService);
+        }
+
+        static RequestParams merge(RequestParams params1, RequestParams params2) {
+            return ImmutableRequestParams.copyOf(params1)
+                    .withNumTransactions(params1.numTransactions() + params2.numTransactions());
+        }
+    }
+
+    @Value.Immutable
+    interface SettableResponse {
+        @Value.Parameter
+        Integer numTransactions();
+
+        @Value.Parameter
+        DisruptorFuture<List<StartIdentifiedAtlasDbTransactionResponse>> startTransactionsFuture();
+
+        static SettableResponse of(
+                int numTransactions, DisruptorFuture<List<StartIdentifiedAtlasDbTransactionResponse>> future) {
+            return ImmutableSettableResponse.of(numTransactions, future);
+        }
+    }
+
+    private static class MultiClientRequestManager implements AutoCloseable {
+        private final Map<Namespace, RequestParams> requestMap;
+
+        public MultiClientRequestManager(Map<Namespace, RequestParams> requestMap) {
+            this.requestMap = requestMap;
+        }
+
+        public boolean requestsPending() {
+            return !requestMap.isEmpty();
+        }
+
+        public void updatePendingStartTransactionsCount(Namespace namespace, int startedTransactionsCount) {
+            Optional<RequestParams> updatedParams = paramsAfterResponse(namespace, startedTransactionsCount);
+            if (updatedParams.isPresent()) {
+                requestMap.put(namespace, updatedParams.get());
+            } else {
+                requestMap.remove(namespace);
+            }
+        }
+
+        private Optional<RequestParams> paramsAfterResponse(Namespace namespace, int startedTransactionsCount) {
+            RequestParams params = requestMap.get(namespace);
+            int numTransactions = params.numTransactions();
+            return startedTransactionsCount < numTransactions
+                    ? Optional.of(RequestParams.of(
+                            numTransactions - startedTransactionsCount, params.cache(), params.lockCleanupService()))
+                    : Optional.empty();
+        }
```

Comment:
Also, I think this is kinda nicer, but double check that it makes sense :)

```suggestion
        public void updatePendingStartTransactionsCount(Namespace namespace, int startedTransactionsCount) {
            requestMap.compute(namespace, (_unused, params) -> paramsAfterResponse(params, startedTransactionsCount));
        }

        private RequestParams paramsAfterResponse(RequestParams params, int startedTransactionsCount) {
            int numTransactions = params.numTransactions();
            if (startedTransactionsCount < numTransactions) {
                return ImmutableRequestParams
                        .builder()
                        .from(params)
                        .numTransactions(numTransactions - startedTransactionsCount)
                        .build();
            }
            return null;
        }
```